### PR TITLE
Improve exception message

### PR DESCRIPTION
**Pull request by VitalyVPinchuk**
For #4615 
Report detailed location of an error using first stack trace element.

> Analyzing /Users/vvp/IdeaProjects/detekt/detekt-core/build/resources/test/cases/Test.kt led to an exception. 
> Location: AnalyzerSpec.kt => io.gitlab.arturbosch.detekt.core.FaultyRule => visitKtFile => 132.
> The original exception message was: Deliberately triggered error.
> Running detekt '1.20.0' on Java '1.8.0_332-b08' on OS 'Mac OS X'
> If the exception message does not help, please feel free to create an issue on our GitHub page.

### CODE REVIEW

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Analyzer.kt [hunk 1]
```diff
@@ -153,8 +153,13 @@ private fun MutableMap<String, List<Finding>>.mergeSmells(other: Map<String, Lis
 }
 
 private fun throwIllegalStateException(file: KtFile, error: Throwable): Nothing {
+    val location = error.stackTrace.firstOrNull()?.let {
+        "${it.fileName} => ${it.className} => ${it.methodName} => ${it.lineNumber}."
+    } ?: "Unknown."
```
Comment (schalkms) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Analyzer.kt:158]: Could we provide a test case that triggers the alternative *unknown* path?

**Review by schalkms — COMMENTED** [review]
Thanks for the PR! I like this addition.

File: detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/AnalyzerSpec.kt [hunk 2]
```diff
@@ -121,14 +127,20 @@ private class MaxLineLength(config: Config, threshold: Int?) : Rule(config) {
     }
 }
 
-private class FaultyRuleSetProvider : RuleSetProvider {
+private class FaultyRuleSetProvider(private val isStackTraceEmpty: Boolean) : RuleSetProvider {
```
[the remark this answers is not in the document]
Comment (VitalyVPinchuk) [reply detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/AnalyzerSpec.kt]: done

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Analyzer.kt [hunk 3]
```diff
@@ -153,8 +153,13 @@ private fun MutableMap<String, List<Finding>>.mergeSmells(other: Map<String, Lis
 }
 
 private fun throwIllegalStateException(file: KtFile, error: Throwable): Nothing {
+    val location = error.stackTrace.firstOrNull()?.let {
+        "${it.fileName} => ${it.className} => ${it.methodName} => ${it.lineNumber}."
```
Comment (BraisGabin) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Analyzer.kt:157]: I would use `firstOrNull()?.toString()`.

So we get the location like this: `io.gitlab.arturbosch.detekt.core.FaultyRule.visitKtFile(AnalyzerSpec.kt:132)`.

Reasons:
- As developers we are used to read locations at stacktraces and they look like that.
- IntelliJ understand that regex and it creates direct links to the code.

**Review by BraisGabin — COMMENTED** [review]
Thank you very much to track this one :). This will help a lot.

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Analyzer.kt (hunk 3 above)
Comment (VitalyVPinchuk) [reply detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/Analyzer.kt]: Agree! That's much better! Thanks!

**Review by BraisGabin — APPROVED** [review]

**Pull request merged**